fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,673 @@
+        # needs pro-rata calculation
+        # First, calculate the number of committees in the current epoch
+        num_slots_in_current_epoch = int((end_slot % SLOTS_PER_EPOCH) + 1)
+        # Next, calculate the number of slots remaining in the current epoch
+        remaining_slots_in_current_epoch = int(SLOTS_PER_EPOCH - num_slots_in_current_epoch)
+        # Then, calculate the number of slots in the previous epoch
+        num_slots_in_previous_epoch = int(SLOTS_PER_EPOCH - (start_slot % SLOTS_PER_EPOCH))
+
+        # Each committee from the previous epoch only contributes a pro-rated weight
+        return Gwei(
+            (num_slots_in_current_epoch * int(committee_weight)) + 
+            (
+                num_slots_in_previous_epoch * remaining_slots_in_current_epoch *
+                int(committee_weight) // SLOTS_PER_EPOCH
+            )
+        )
+```
+
+#### `is_one_confirmed`
+
+```python
+def is_one_confirmed(store: Store, confirmation_byzantine_threshold: int, block_root: Root) -> bool:
+    current_slot = get_current_slot(store)
+    block = store.blocks[block_root]
+    parent_block = store.blocks[block.parent_root]
+    support = int(get_weight(store, block_root))
+    justified_state = store.checkpoint_states[store.justified_checkpoint]
+    maximum_support = int(
+        get_committee_weight_between_slots(justified_state, Slot(parent_block.slot + 1), current_slot)
+    )
+    proposer_score = int(get_proposer_score(store))
+
+    """
+    Returns whether the following condition is true using only integer arithmetic
+    support / maximum_support >
+    0.5 * (1 + proposer_score / maximum_support) + confirmation_byzantine_threshold / 100
+    """
+
+    return (
+        100 * support > 
+        50 * maximum_support + 50 * proposer_score + confirmation_byzantine_threshold * maximum_support 
+    )
+```
+
+#### `is_lmd_confirmed`
+
+```python
+def is_lmd_confirmed(store: Store, confirmation_byzantine_threshold: int, block_root: Root) -> bool:
+    if block_root == store.finalized_checkpoint.root:
+        return True
+    else:
+        block = store.blocks[block_root]
+        finalized_block = store.blocks[store.finalized_checkpoint.root]
+        if block.slot <= finalized_block.slot:
+            # This block is not in the finalized chain.
+            return False
+        else:
+            # Check is_one_confirmed for this block and is_lmd_confirmed for the preceding chain.
+            return (
+                is_one_confirmed(store, confirmation_byzantine_threshold, block_root) and
+                is_lmd_confirmed(store, confirmation_byzantine_threshold, block.parent_root)
+            )
+```
+
+#### `get_total_active_balance_for_block_root`
+
+```python
+def get_total_active_balance_for_block_root(store: Store, block_root: Root) -> Gwei:
+    assert block_root in store.block_states
+
+    # Alternatively, fetch the state from the block's checkpoint
```

Comment:
I'm not sure what we mean by "Alternatively" here. As there is just one potential code path

Reply:
Yeah. This is just a leftover from previous code refactoring. Fixed now.